### PR TITLE
Add support for public key authentication and encryption

### PR DESCRIPTION
**Pull request by rubidium42**
With OpenTTD/OpenTTD#11878 a new and more secure way of doing authentication is added. To test these changes in OpenTTD, I needed a client and I used this one so why not try to upstream these changes. This branch has not been merged, and if it will be merged it is going to be slated for version 15 of OpenTTD so there is no real hurry yet.

The concept of the new authentication is that there is a private-public key exchange using X25519. Essentially the maths works out that you can generate a shared secret by knowing each other's public key, your own private key and some random (unique) data.
With this shared secret a new encryption key is derived, which is done by hashing some values together. With this encryption key a message is encrypted and sent back to the server, which performs the same steps and checks whether it can decrypt the message and whether the signature is valid. If that works out some further checks can be done at the server side.

Currently there are two methods:
* password authenticated key exchange; here the password is one of the values that get hashed together, and as such is used to generate the encryption key. If you pass the wrong password, your encryption key will be different from the server's encryption key, and anything the client encrypts cannot be decrypted and validated by the server. This means that it can check for knowledge of the password without actually sending the password.
* authorized keys; here the user generates a secret and public key. The user configures the client to use the secret key, and sends the public key to the server owner to put in their list of allowed (public) keys. Due to the key exchange only working with one of the two secret keys and both public keys, the client proves that it is the owner of the secret key associated with the public key and the server checks whether that public key is authorized.

The latter allows more fine-grained control to who may login at any time, as each client is essentially an unique entity. 
The biggest caveat is that the current join packet contains the password, and we do not want to send the password any more over the wire, so a new join packet has been introduced. In this packet the client can tell which authentication methods it support, e.g. if the user does not pass a password, then it can leave out that method.

The server will choose the "best" method for logging in, which favours authorized keys if that's supported by the client. When that fails, it will fall back to the password variant if the client supports that and the server has a password set.

After authentication using the new join packet the data on the network will be encrypted using the encryption keys generated by the last key exchange.
With the old join packet everything remains as it is, so password without encryption over the wire and no encryption for the data. The default for OpenTTD will be to disable that insecure version, and I added a commit to change the default in this project to the secure join method (the last one). Maybe that commit needs to be merged later, when 15 is near and clients should use that as for the current OpenTTD versions the secure join method simply will not work, and it will forcefully close the connection.

For this change there is a requirement on some cryptography library. The problem is that the "standard" ones do not have all the functionality that is required, so you'd need to pick and choose.
To prevent any kind of inconsistency issues with the different implementations, I have looked into Python wrappers around the library that OpenTTD uses: Monocypher. That exists in the form of pymonocypher.
The "only" problem as of writing is that the current version is an old version, which behaves slightly different from the newer version used in OpenTTD. Next to that the stuff for streaming encryption is not exposed yet, so I made a few pull requests to get a newer version and expose that: jetperch/pymonocypher#8.

That does mean that this PR cannot be merged yet, because it depends on an unreleased version of pymonocypher and because the feature has not been merged into OpenTTD yet. But it gives you some time to gather your thoughts on this PR.
I hope I didn't butcher this library and Python too much ;)

Since the crypto.py part might be useful to be used for other projects talking to OpenTTD, I'll hereby also license it under the GNU Lesser General Public License version 2.1 (LGPL 2.1).

### CODE REVIEW

File: libottdadmin2/packets/__init__.py
Comment (Xaroth) [inline libottdadmin2/packets/__init__.py:13]: Should this not be something like `AdminJoinSecure`, since this only applies to secure connections?

Comment (rubidium42) [reply libottdadmin2/packets/__init__.py]: Good point. It was already renamed in OpenTTD, but I never fixed that here.

**Comment by rubidium42** [comment]
There has been a release of pymonocypher with the required functionality, so you'd just need to `pip install pymonocypher` to get the required dependency installed.

Now it's waiting for the feature to be merged into OpenTTD before merging here I'd reckon.

**Review by Xaroth — APPROVED** [review]
Let me know (pinging on discord works) when this merges into main on OpenTTD, and I'll get this merged in as well.

**Comment by rubidium42** [comment]
The functionality was just merged into master in OpenTTD.

**Pull request merged**
